habitat/core/dataset.py

```diff
@@ -361,3 +373,47 @@ def _shuffle_iterator(self) -> None:
             episodes = list(self._iterator)
             random.shuffle(episodes)
             self._iterator = iter(episodes)
+
+    def step_taken(self):
+        self._step_count += 1
+
+    def _set_shuffle_intervals(self):
+        if self.max_scene_repetition_episodes > 0:
+            self._max_rep_episode = random.randint(
+                int(0.8 * self.max_scene_repetition_episodes),
+                int(1.2 * self.max_scene_repetition_episodes),
+            )
+        else:
+            self._max_rep_episode = None
+
+        if self.max_scene_repetition_steps > 0:
+            self._max_rep_step = random.randint(
+                int(0.8 * self.max_scene_repetition_steps),
+                int(1.2 * self.max_scene_repetition_steps),
+            )
+        else:
+            self._max_rep_step = None
+
+    def _switch_scene_if(self):
+        do_switch = False
+        self._rep_count += 1
+
+        # Shuffle if a scene has been selected more than _max_rep_episode times in a row
+        if (
+            self._max_rep_episode is not None
+            and self._rep_count >= self._max_rep_episode
+        ):
+            do_switch = True
+
+        # Shuffle if a scene has been used for more than _max_rep_step steps in a row
+        if (
+            self._max_rep_step is not None
+            and self._step_count >= self._max_rep_step
+        ):
+            do_switch = True
+
+        if do_switch:
+            self._shuffle_iterator()
```

Comment:
Should we check for `self.shuffle == True`, here when we shuffling.

Reply:
No.  The `_shuffle_iterator` is used both to initiate a scene switch and to just the shuffle the episodes.  `self.shuffle` decides whether or not to shuffle the episode order on cycle or on load.

Reply:
That's strange as there was no option to enable `cycling` without episode `shuffling`.

Reply:
The cycle logic doesn't really on the shuffle